### ISSUE TITLE
Issues with using Firefox addon VideoDownloadHelper's "companion app"

### ISSUE THREAD

**Issue by bobafetthotmail** [netblue30/firejail#2987]
The popular extension VideoDownloadHelper (for both Firefox and Chrome) relies on an additional application https://www.downloadhelper.net/install-coapp?browser=firefox to be able to download high-resolution media.

I've extracted the archive in (HOME)/dwhelper/net.downloadhelper.coapp/ and ran the "user installation" (which is just generating a config file that links to the application, as I documented here https://github.com/mi-g/vdhcoapp/issues/47#issuecomment-537663897 )

These are the contents of that folder
```

ls -lR /home/alby/dwhelper/net.downloadhelper.coapp/
/home/alby/dwhelper/net.downloadhelper.coapp/:
total 28
drwxr-xr-x 1 alby users    82 28 lug  2018 bin
-rw-r--r-- 1 alby users   456  4 lug  2018 config.json
drwxr-xr-x 1 alby users    10 28 lug  2018 converter
-rw-r--r-- 1 alby users 18092 25 ott  2017 LICENSE.txt
-rw-r--r-- 1 alby users   571 25 ott  2017 README.txt

/home/alby/dwhelper/net.downloadhelper.coapp/bin:
total 56152
-rwxr-xr-x 1 alby users 57467401  4 lug  2018 net.downloadhelper.coapp-linux-64
-rwxr-xr-x 1 alby users    25047  4 lug  2018 xdg-open

/home/alby/dwhelper/net.downloadhelper.coapp/converter:
total 0
drwxr-xr-x 1 alby users 10  4 lug  2018 build

/home/alby/dwhelper/net.downloadhelper.coapp/converter/build:
total 0
drwxr-xr-x 1 alby users 4  4 lug  2018 linux

/home/alby/dwhelper/net.downloadhelper.coapp/converter/build/linux:
total 0
drwxr-xr-x 1 alby users 1136  4 lug  2018 64

/home/alby/dwhelper/net.downloadhelper.coapp/converter/build/linux/64:
total 34224
-rwxr-xr-x 1 alby users   255208 24 mag  2018 ffmpeg
-rwxr-xr-x 1 alby users   127656 24 mag  2018 ffplay
-rwxr-xr-x 1 alby users   147616 24 mag  2018 ffprobe
-rwxr-xr-x 1 alby users 13381168 24 mag  2018 libavcodec.so.58
-rwxr-xr-x 1 alby users    96512 24 mag  2018 libavdevice.so.58
-rwxr-xr-x 1 alby users  2730520 24 mag  2018 libavfilter.so.7
-rwxr-xr-x 1 alby users  2391392 24 mag  2018 libavformat.so.58
-rwxr-xr-x 1 alby users   129472 24 mag  2018 libavresample.so.4
-rwxr-xr-x 1 alby users   416528 24 mag  2018 libavutil.so.56
-rwxr-xr-x 1 alby users    66984 24 mag  2018 libbz2.so.1.0
-rwxr-xr-x 1 alby users   585392 24 mag  2018 libmp3lame.so.0
-rwxr-xr-x 1 alby users    44184 24 mag  2018 libnuma.so.1
-rwxr-xr-x 1 alby users    26888 24 mag  2018 libogg.so.0
-rwxr-xr-x 1 alby users   170104 24 mag  2018 libopencore-amrnb.so.0
-rwxr-xr-x 1 alby users    79920 24 mag  2018 libopencore-amrwb.so.0
-rw-r--r-- 1 alby users   307304 24 mag  2018 libopenjp2.so.7
-rwxr-xr-x 1 alby users   330136 24 mag  2018 libopus.so.0
-rwxr-xr-x 1 alby users   525360 24 mag  2018 liborc-0.4.so.0
-rwxr-xr-x 1 alby users    31152 24 mag  2018 liborc-test-0.4.so.0
-rwxr-xr-x 1 alby users   112912 24 mag  2018 libpostproc.so.55
-rwxr-xr-x 1 alby users   117152 24 mag  2018 libswresample.so.3
-rwxr-xr-x 1 alby users   526736 24 mag  2018 libswscale.so.5
-rwxr-xr-x 1 alby users    96656 24 mag  2018 libtheoradec.so.1
-rwxr-xr-x 1 alby users   260496 24 mag  2018 libtheoraenc.so.1
-rwxr-xr-x 1 alby users   305472 24 mag  2018 libtheora.so.0
-rwxr-xr-x 1 alby users   107248 24 mag  2018 libvo-amrwbenc.so.0
-rwxr-xr-x 1 alby users   706768 24 mag  2018 libvorbisenc.so.2
-rwxr-xr-x 1 alby users    39400 24 mag  2018 libvorbisfile.so.3
-rwxr-xr-x 1 alby users   232232 24 mag  2018 libvorbis.so.0
-rwxr-xr-x 1 alby users  3672648 24 mag  2018 libvpx.so.5
-rwxr-xr-x 1 alby users   215544 24 mag  2018 libwebpdecoder.so.3
-rwxr-xr-x 1 alby users    18640 24 mag  2018 libwebpdemux.so.2
-rwxr-xr-x 1 alby users    43360 24 mag  2018 libwebpmux.so.3
-rwxr-xr-x 1 alby users   416440 24 mag  2018 libwebp.so.7
-rwxr-xr-x 1 alby users   973784 24 mag  2018 libx264.so.152
-rwxr-xr-x 1 alby users  4679048 24 mag  2018 libx265.so.130
-rw-r--r-- 1 alby users   494792 24 mag  2018 libxvidcore.so.4
-rwxr-xr-x 1 alby users   113152 24 mag  2018 libz.so.1
```

and then tried adding things to the **/etc/firejail/firefox-common-addons.inc**
until I ended up with these additions
```
noblacklist ${HOME}/dwhelper
mkdir ${HOME}/dwhelper
whitelist ${HOME}/dwhelper

mkdir ${HOME}/dwhelper/net.downloadhelper.coapp/
mkdir ${HOME}/dwhelper/net.downloadhelper.coapp/bin
mkdir ${HOME}/dwhelper/net.downloadhelper.coapp/converter
whitelist ${HOME}/dwhelper/net.downloadhelper.coapp/
whitelist ${HOME}/dwhelper/net.downloadhelper.coapp/bin
whitelist ${HOME}/dwhelper/net.downloadhelper.coapp/converter
```

And each time I restarted Firefox. But it does not seem to work, the addon can't find its "companion app". Anyone has some advice?

EDIT: If I place the "companion app" in the /downloads folder (which is whitelisted and accessible by default) I get 

`Failed to execute command "/home/alby/Downloads/net.downloadhelper.coapp/bin/net.downloadhelper.coapp-linux-64"`

as error message in the addon when I ask it to detect the presence of its "companion app"

**Comment by rusty-snake** [comment]
1. Do you include the firefox-common-addons.inc profile?
2. No trailing slashes for `mkdir`/`whitelist`
3. `whitelist ${HOME}/dwhelper` should be enought.
4. Set `browser-allow-drm yes` in `firejail.config`

**Comment by rusty-snake** [comment]
`firejail --whitelist='${HOME}/dwhelper --ignore='noexec ${HOME}' firefox`

**Comment by bobafetthotmail** [comment]
1. I thought it was included, but you are right, it's not included. I have now uncommented its include line in **firefox-common.inc**
2 and 3. ok removing other lines
4. done this too

As for your second post, I can't run firefox (or thunderbird) from commandline. Other applications (say filezilla) can be called from command line like that.

I converted the "ignore noexec" into configuration.

Now I have added only
`whitelist ${HOME}/dwhelper
ignore noexec ${HOME}/dwhelper`
to  firefox-common-addons.inc

And something is different. Instead of giving "unknown issue" I get

`Checking companion app returned: Failed to execute command "/home/alby/dwhelper/net.downloadhelper.coapp/bin/net.downloadhelper.coapp-linux-64"`

So closer, but still no cigar.

**Comment by rusty-snake** [comment]
`ignore noexec ${HOME}/dwhelper` -> `ignore noexec ${HOME}`
which firejail version?

BTW: `ignore noexec ${HOME}` and `browser-allow-drm yes` are the same.
`

**Comment by bobafetthotmail** [comment]
Changed to `ignore noexec ${HOME}`, still the same error as above.

I'm running OpenSUSE Tumbleweed, and latest Firejail release, from what I can tell
```
firejail --version
firejail version 0.9.60

Compile time support:
        - AppArmor support is enabled
        - AppImage support is enabled
        - chroot support is enabled
        - file and directory whitelisting support is enabled
        - file transfer support is enabled
        - networking support is enabled
        - overlayfs support is enabled
        - private-home support is enabled
        - seccomp-bpf support is enabled
        - user namespace support is enabled
        - X11 sandboxing support is enabled
```

**Comment by rusty-snake** [comment]
NOTE: `whitelist ${HOME}/dwhelper` should already in firefox-common-addons.inc

Does `net.downloadhelper.coapp-linux-64` need an interpreter like python? or is it an ELF file?

**Comment by bobafetthotmail** [comment]
Yes it shows as ELF if I open it with a hex editor. 
The application https://github.com/mi-g/vdhcoapp is actually javascript and very light, so I think that's probably an all-in-one javascript runtime + dependencies + application.
 
It seems to be "compiled" with npm (Node JS), only other application it needs is the (static? I don't know) ffmpeg it ships in its own folder
`/home/alby/dwhelper/net.downloadhelper.coapp/converter/build/linux/64/`

**Comment by rusty-snake** [comment]
Can it be execute with `firejail --profile=firefox /home/alby/dwhelper/net.downloadhelper.coapp/bin/net.downloadhelper.coapp-linux-64`?

**Comment by bobafetthotmail** [comment]
It seems so. If started with no arguments it sits there not printing nothing on console until i Ctrl+C and then it closes, like it does if I invoke it without firejail.

```
firejail --profile=firefox /home/alby/dwhelper/net.downloadhelper.coapp/bin/net.downloadhelper.coapp-linux-64
Reading profile /etc/firejail/firefox.profile
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/firefox-common-addons.inc
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Parent pid 21567, child pid 21568
Warning: cleaning all supplementary groups
Warning: cleaning all supplementary groups
Warning: cleaning all supplementary groups
Warning: cleaning all supplementary groups
Post-exec seccomp protector enabled
Seccomp list in: @clock,@cpu-emulation,@debug,@module,@obsolete,@raw-io,@reboot,@resources,@swap,acct,add_key,bpf,fanotify_init,io_cancel,io_destroy,io_getevents,io_setup,io_submit,ioprio_set,kcmp,keyctl,mount,name_to_handle_at,nfsservctl,ni_syscall,open_by_handle_at,personality,pivot_root,process_vm_readv,ptrace,remap_file_pages,request_key,setdomainname,sethostname,syslog,umount,umount2,userfaultfd,vhangup,vmsplice, check list: @default-keep, prelist: adjtimex,clock_adjtime,clock_settime,settimeofday,modify_ldt,lookup_dcookie,perf_event_open,process_vm_writev,delete_module,finit_module,init_module,_sysctl,afs_syscall,create_module,get_kernel_syms,getpmsg,putpmsg,query_module,security,sysfs,tuxcall,uselib,ustat,vserver,ioperm,iopl,kexec_load,kexec_file_load,reboot,set_mempolicy,migrate_pages,move_pages,mbind,swapon,swapoff,acct,add_key,bpf,fanotify_init,io_cancel,io_destroy,io_getevents,io_setup,io_submit,ioprio_set,kcmp,keyctl,mount,name_to_handle_at,nfsservctl,open_by_handle_at,personality,pivot_root,process_vm_readv,ptrace,remap_file_pages,request_key,setdomainname,sethostname,syslog,umount2,userfaultfd,vhangup,vmsplice,
Child process initialized in 129.31 ms
^C
Parent received signal 2, shutting down the child process...

Child received signal 2, shutting down the sandbox...

Parent is shutting down, bye...
```

This is with the "install" command so it can actually answer something on the terminal, the 

"VdhCoApp: VdhCoApp is ready to be used"

 is the application output.
```
firejail --profile=firefox /home/alby/dwhelper/net.downloadhelper.coapp/bin/net.downloadhelper.coapp-linux-64 install --user
Reading profile /etc/firejail/firefox.profile
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/firefox-common-addons.inc
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Parent pid 21597, child pid 21598
Warning: cleaning all supplementary groups
Warning: cleaning all supplementary groups
Warning: cleaning all supplementary groups
Warning: cleaning all supplementary groups
Post-exec seccomp protector enabled
Seccomp list in: @clock,@cpu-emulation,@debug,@module,@obsolete,@raw-io,@reboot,@resources,@swap,acct,add_key,bpf,fanotify_init,io_cancel,io_destroy,io_getevents,io_setup,io_submit,ioprio_set,kcmp,keyctl,mount,name_to_handle_at,nfsservctl,ni_syscall,open_by_handle_at,personality,pivot_root,process_vm_readv,ptrace,remap_file_pages,request_key,setdomainname,sethostname,syslog,umount,umount2,userfaultfd,vhangup,vmsplice, check list: @default-keep, prelist: adjtimex,clock_adjtime,clock_settime,settimeofday,modify_ldt,lookup_dcookie,perf_event_open,process_vm_writev,delete_module,finit_module,init_module,_sysctl,afs_syscall,create_module,get_kernel_syms,getpmsg,putpmsg,query_module,security,sysfs,tuxcall,uselib,ustat,vserver,ioperm,iopl,kexec_load,kexec_file_load,reboot,set_mempolicy,migrate_pages,move_pages,mbind,swapon,swapoff,acct,add_key,bpf,fanotify_init,io_cancel,io_destroy,io_getevents,io_setup,io_submit,ioprio_set,kcmp,keyctl,mount,name_to_handle_at,nfsservctl,open_by_handle_at,personality,pivot_root,process_vm_readv,ptrace,remap_file_pages,request_key,setdomainname,sethostname,syslog,umount2,userfaultfd,vhangup,vmsplice,
Child process initialized in 132.44 ms
VdhCoApp: VdhCoApp is ready to be used

Parent is shutting down, bye...
```

**Comment by rusty-snake** [comment]
Hmm, the profile allows executeing this file, but VDH can't execute it. Can you check if it works with `firejail --noprofile firefox` so we can see if we need to tune the profile or the issues is deeper.

**Comment by Vincent43** [comment]
It may be blocked by apparmor. Please try `ignore apparmor` in `firefox.local`.

**Comment by bobafetthotmail** [comment]
tried with `ignore apparmor` and no difference.

`firejail --noprofile firefox` works and the addon can detect its "companion app".

**Comment by pizzadude** [comment]
Can you try this? This works for me:

noblacklist ~/net.downloadhelper.coapp
whitelist ~/net.downloadhelper.coapp

(Change the path to where you have it installed)
Also make sure you do this: ~/net.downloadhelper.coapp/bin/net.downloadhelper.coapp-linux-64 install --user

It works since forever in firejail stable, and git versions (currently running from latest git)

**Comment by bobafetthotmail** [comment]
Hm, what you did isn't different from what I also tried (I just edited the main config files, that would be overwritten on update).

I just tried and your method does not work for me.

Maybe this is a distro default configuration of something? What is your distro?

**Comment by rusty-snake** [comment]
@bobafetthotmail have you edited some thing else? e.g. `private-bin`

**Comment by pizzadude** [comment]
@bobafetthotmail My distro is Fedora 31. I am running firejail from git.

**Comment by bobafetthotmail** [comment]
No I didn't edit anything more than what I said here or you recommended.

But I tried adding `ignore apparmor` to @pizzadude firefox.local because of what I've seen below, and now it is working.

Any way forward?

> @bobafetthotmail My distro is Fedora 31. I am running firejail from git.

Hm, Fedora uses SELinux and not AppArmor. The AppArmor config is getting in the way here.

In Yast (graphical system configuration manager for OpenSUSE) I see that AppArmor has an "audit module" and if I ask it to read the logs it does ask me to do something about the problem at hand.

```
Execute: /home/alby/dwhelper/net.downloadhelper.coapp/bin/net.downloadhelper.coapp-linux-64
Severity 0:
```
And then a row of buttons:

`Inherit | Child | Profile | Named | Unconfined | X ix On | Deny | Abort | Finish`

I don't know what I should answer, or if this is even supposed to happen (isn't firejail supposed to deal with apparmor on its own?) .

I've answered "Abort" and this closes the window without taking any decision.

Any ideas?
